ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Counting.hs

```diff
@@ -361,6 +362,9 @@ modifyMV (MVector mv) f (Count i)   = MV.modify mv f i
 readV :: MV.Unbox a => Vector base elem a -> Index base elem -> a
 readV (Vector v) (Count i) = v V.! i
 
+growMV :: MV.Unbox a => MVector base elem s a -> Size base elem -> ST s (MVector base elem s a)
```

Comment:
Maybe we can rename it to `unsafeGrowMV`.

Reply:
My instincts say that'd merely be deferring the pain to our future selves :)

I think any alternative that involves less surprise would not extend the already-generated alternative chain. In other words, it'd have to involved an additional output.

```
newtype Addendum = Addendum (V.Vector S)

uniformAdversarialChain ::
  forall g base hon adv.
     R.RandomGen g
  => Maybe Asc   -- ^ 'Nothing' means @1@
  -> CheckedAdversarialRecipe base hon adv
  -> g
  -> (ChainSchema base adv, Addendum)
```

And then we'd have a mitigating function along the lines of the following:

```
data Bundle base hon adv = Bundle
    (CheckedHonestRecipe base hon)
    (CheckedAdversrialRecipe base adv)
    (ChainSchema base hon)
    (ChainSchema base adv)

extendAlternativeSchema ::  Bundle base hon adv -> Addendum -> (forall base' hon' adv'. Bundle base' hon' adv' -> ??? -> ans) -> ans
```

where the `???` is whatever set of `Contains` values would embed `base` in `base'`, `hon` in `hon`' ,and `adv` in `adv'`---all of which result from simply extending by the length of the `Addendum`. For the honest chain, it'd be safest to make _all_ of the new slots active: that preserves CG and moreover these slots are already irrelevant to the alternative schema since the alternative schema is already accelerating. (TODO I'm not certain about that second claim; what do you think?)

I'm anticipating that the `Arbitrary` instances for test setups would simply call this as it's last step: almost all of our code would never see the initial set of type variables 🤞.

That seems like the only "responsible" approach to me, at least at this brainstorming stage. IE, the approach that still respects the type-level safety while adding this new "post-processing" step we've conceived. (Since that step involves adding new slots to the end of time.)

Ah ha! That reminds me: this PR should also rename `Len` to `LenAtLeast` or similar, right?

Reply:
> Ah ha! That reminds me: this PR should also rename Len to LenAtLeast or similar, right?

The Len still defines accurately the len of honest schemas, which is what it seems to control based on its location. `LenAtLeast` is going to be confusing for honest schemas. I'd leave this aspect for a separate PR.

Reply:
Alrighty. Then I'd say we at least need a comment explaining something along the lines of the adversarial schema might be longer than `Len`.

Reply:
Added a comment to `uniformAdversarialChain`.

Reply:
I'm replying to https://github.com/input-output-hk/ouroboros-consensus/pull/825#issuecomment-1852477990 here, since I think it's the same issue and I want to keep the discussion threaded.

Recall these lines.

```
data ChainSchema base inner =
    ChainSchema
        !(C.Contains SlotE base inner)
        !(C.Vector inner SlotE S)
```

```
data AdversarialRecipe base hon =
    AdversarialRecipe {
        arHonest :: !(ChainSchema base hon)
      ,
        {- snipped -}
      }
```

```
checkAdversarialRecipe ::
  forall base hon.
     AdversarialRecipe base hon
  -> Exn.Except
         NoSuchAdversarialChainSchema
         (SomeCheckedAdversarialRecipe base hon)
```

```
data CheckedAdversarialRecipe base hon adv =
    UnsafeCheckedAdversarialRecipe {
        {- snipped -}
      ,
        carWin    :: !(C.Contains SlotE hon adv)
      }
```

```
uniformAdversarialChain ::
  forall g base hon adv.
     R.RandomGen g
  => Maybe Asc   -- ^ 'Nothing' means @1@
  -> CheckedAdversarialRecipe base hon adv
  -> g
  -> ChainSchema base adv
```

For example, altogether, these signatures' semantics (specifically `carWin`) imply that the adversarial schema is not longer than the honest schema.

Reply:
> give up on using type checking to assert sizes and windows for adversarial schemas.

I think that's fine to do _downstream_ of the chain generators, but not inside of them. So I'd suggest my `Addendum` data type to `uniformAdversarialChain`, but then some function simpler than my `extendAlternativeSchema` function could be used as the outermost entrypoint to the adversarial chain generator?

Reply:
@tek suggested something like this in a call today
```Haskell
data GrownMVector base elem s a = forall w.                                                                                                                                                                                                  
    GrownMVector (Contains elem w base) (MVector w elem s a)                                                                                                                                                                                 
                                                                                                                                                                                                                                             
growMV :: MV.Unbox a => MVector base elem s a -> Size base elem -> ST s (GrownMVector base elem s a)                                                                                                                                         
growMV (MVector mv) (Count n) = do                                                                                                                                                                                                           
    mv' <- MV.grow mv n                                                                                                                                                                                                                      
    pure $ GrownMVector                                                                                                                                                                                                                      
      (UnsafeContains (Count 0) (Count n))                                                                                                                                                                                     
      (MVector mv')  
```

I think this could be fine for now.

Reply:
I explored this idea in the last batch of commits. Let me know how you want to proceed.

> I think that's fine to do downstream of the chain generators, but not inside of them.

If we go this other way, I would give the following type to `uniformAdversarialChain`.
```Haskell
uniformAdversarialChain ::                                                                                                                                                                                                                   
  forall g base hon. -- adv.                                                                                                                                                                                                                     
     R.RandomGen g                                                                                                                                                                                                                           
  => Maybe Asc
  -> CheckedAdversarialRecipe base hon -- adv
  -> g
  -> Vector S
```

Internally, the function can work with the indexed types, but callers do not need to deal with them. `CheckedAdversarialRecipe` would likely need to be adapted too, to account for the fact that we are removing the `adv` type parameter.

In the long run, I feel like we need to aim for simplifying the type checking that we do in these generators. My impression is that it is making the code too difficult to change.

> So I'd suggest my Addendum data type to uniformAdversarialChain, but then some function simpler than my extendAlternativeSchema function could be used as the outermost entrypoint to the adversarial chain generator?

While we might do this in the end, at the moment I feel like the other options are better. Happy to discuss it some more.